### PR TITLE
Commiting Test Folders

### PR DESCRIPTION
**Pull request by BACStrange**
Adding tests for Sanity Check, False Metals Check and Nearest Neighbor Distance parsing test

### CODE REVIEW

File: tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR [hunk 1]
```diff
@@ -0,0 +1,17 @@
+SYSTEM = r-GeO2 ENCUT convergence test
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR]: The system name appears to be incorrect here. Since it is not required for a test input file, it could be removed.

File: tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+NSW    = 0
+IBRION = -1
+ISMEAR = 0
+SIGMA  = 0.05
+LREAL=.FALSE.
+ALGO   = Normal
+LWAVE  = .FALSE.
+LCHARG = .FALSE.
+LASPH  = .TRUE.
+NPAR   = 4
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR]: I am not sure whether NPAR=4 is suitable for a test, especially on systems where 4 cores may not be available. It might be better to remove it.

File: tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/KPOINTS [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Automatic mesh
+0
+Gamma
+15 15 15
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/KPOINTS]: I would suggest reconsidering this k-point mesh, as it may be too dense for a test case. A denser grid also requires reference `EIGENVAL` files that can become unnecessarily large.
Would it make sense to reduce it to a minimal grid, such as `2x2x2`?

File: tests/README.txt
Comment (ronaldorpela) [inline tests/README.txt:1]: If there is no reason against it, I would recommend having this README file in markdown format

File: tests/README.txt [hunk 4]
```diff
@@ -0,0 +1,27 @@
+VASP Test Setup Instructions
+
+To run the VASP tests, you must provide the required POTCAR files for the following elements:
```
Comment (ronaldorpela) [inline tests/README.txt]:
```suggestion
To run the VASP tests, the required POTCAR files for the following elements must be provided:
```

File: tests/README.txt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+Ag
+C
+F
+S
+Sb
+
+Place these files in the directory:
```
Comment (ronaldorpela) [inline tests/README.txt]:
```suggestion
Place these files in the following directory:
```

File: tests/README.txt [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+test_farm/VESP/POTFILES
```
Comment (ronaldorpela) [inline tests/README.txt]:
```suggestion
test_farm/VASP/POTFILES
```

File: tests/README.txt [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+Each file must be named using the following convention:
```
Comment (ronaldorpela) [inline tests/README.txt]:
```suggestion
Each file must follow this naming scheme:
```

File: tests/README.txt [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+where <element> is written in lowercase. For example:
+
+POTCAR.ag
+POTCAR.c
+POTCAR.f
+POTCAR.s
+POTCAR.sb
+
+Ensure that all files are correctly named and located before running the tests.
```
Comment (ronaldorpela) [inline tests/README.txt]: In my opinion, this line seems redundant and could be removed

File: tests/run_tests.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
```
Comment (ronaldorpela) [inline tests/run_tests.py]: The shebang line
```python
#!/usr/bin/env python3
```
is only useful if the script is intended to be:
- executed directly (./script.py), and
- used consistently that way across the project

If the project does not follow this pattern, then including it can be unnecessary or inconsistent.

File: tests/run_tests.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+"""
+run_tests.py
+------------
+Reads the test registry from mhtests.yml and runs each test sequentially.
```
Comment (ronaldorpela) [inline tests/run_tests.py]: To be consistent with the rest of the documentation, imperative should be used

Reads -> Read

Another point: I was not able to find the mhtests.yml file

File: tests/run_tests.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+For each test it:
+1. Resolves the absolute path of the test script
+2. Writes that path to the minushalf.yaml file
+3. Runs 'minushalf execute' inside the test folder
+4. Reports pass or fail
+
+Usage:
+    python run_tests.py
+
+To add a new test, just add an entry to mhtests.yml — no changes needed here.
```
Comment (ronaldorpela) [inline tests/run_tests.py]:
```suggestion
For each test, the script performs the following steps:
1. Resolves the absolute path of the test script
2. Writes this path to the minushalf.yaml file
3. Runs 'minushalf execute' within the test directory
4. Reports whether the test passes or fails

Usage:
    python run_tests.py

To add a new test, simply include a new entry in mhtests.yml. No changes to this script are required.
```

File: tests/testconfig.yml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  - name: "3_nnd"
+    description: "Tests for OUTCAR reading bugs"
+    folder: "test_farm/VASP/3_nnd"
+    script: "nndcheck.py"
+
+  # To add a new test, just copy one of the blocks above and fill in the fields:
+  # - name:        unique identifier for the test
+  # - description: human readable description of what the test does
+  # - folder:      path to the test folder relative to the tests/ directory
+  # - script:      name of the mock VASP python script inside that folder
```
Comment (ronaldorpela) [inline tests/testconfig.yml:6]: I consider it a better idea to move the documentation to the beginning of this file

File: tests/testconfig.yml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+tests:
```
Comment (ronaldorpela) [inline tests/testconfig.yml]: Quick question: is the test keyword necessary here? Since the file already lists tests, it feels a bit redundant from my perspective, but I may be missing some context. Could you clarify its role?

File: tests/testconfig.yml [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+  - name: "1_carbon"
+    description: "Sanity check - basic carbon calculation"
+    folder: "test_farm/VASP/1_carbon"
```
Comment (ronaldorpela) [inline tests/testconfig.yml]: Just a thought: if everything is already under test_farm, could we treat it as the base directory and avoid repeating it throughout? It might make the paths a bit cleaner.

File: tests/run_tests.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+# ─────────────────────────────────────────────────────────────────────────────
+# MAIN
+# ─────────────────────────────────────────────────────────────────────────────
+
+if __name__ == "__main__":
+    print(f"\n{'='*60}")
+    print(f"  MHTEST - Minushalf Test Runner")
+    print(f"  Tests root : {TESTS_ROOT}")
+    print(f"  Registry   : {REGISTRY_FILE}")
```
Comment (ronaldorpela) [inline tests/run_tests.py]: I might be missing the point, but what is the motivation for having these two variables as global? I think, local variables might be a better choice.

File: tests/run_tests.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        print(f"\n[FAIL] ❌ {name} failed with return code {result.returncode}.")
+        return False
+
+
+# ─────────────────────────────────────────────────────────────────────────────
+# MAIN
+# ─────────────────────────────────────────────────────────────────────────────
+
+if __name__ == "__main__":
+    print(f"\n{'='*60}")
```
Comment (ronaldorpela) [inline tests/run_tests.py]: If this is repeated along the code, it might be beneficial to have it as an internal (formatting) function.

File: tests/run_tests.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+        print(f"[ERROR] Registry file not found: {REGISTRY_FILE}")
+        sys.exit(1)
+
+    with open(REGISTRY_FILE, "r") as f:
+        data = yaml.safe_load(f)
+
+    return data.get("tests", [])
+
+
+def run_test(test: dict):
```
Comment (ronaldorpela) [inline tests/run_tests.py]: Whenever possible, it is good practice to document the return type as well

File: tests/run_tests.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+    # Verify the test folder exists
+    if not folder.exists():
+        print(f"[ERROR] Test folder not found: {folder}")
+        return False
+
+    # Verify the test script exists
+    if not script.exists():
+        print(f"[ERROR] Test script not found: {script}")
+        return False
```
Comment (ronaldorpela) [inline tests/run_tests.py]: From my perspective, it would be preferable to raise an exception if any of these conditions happen. The same would apply to the presence of minushalf.yml.

File: tests/run_tests.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+
+    ensure_dir(carbon_input_dir)
+    ensure_dir(carbon_mh_dir)
+
+    potcar_c = os.path.join(potfiles_dir, "POTCAR.c")
+
+    copy_file(potcar_c, os.path.join(carbon_mh_dir, "POTCAR.c"))
+    copy_file(potcar_c, os.path.join(carbon_input_dir, "POTCAR"))
+
+    # === 2. Metal ===
```
Comment (ronaldorpela) [inline tests/run_tests.py]: It looks like this part repeats what was done for the first test. Would it make sense to refactor this into a reusable helper? That way, adding new tests to testconfig.yml wouldn’t require duplicating the logic.

File: tests/run_tests.py [hunk 20]
```diff
@@ -0,0 +1,218 @@
...
+    carbon_input_dir = os.path.join(base_dir, "1_carbon", "INPUTS", "cut_2.dummy")
+    carbon_mh_dir = os.path.join(base_dir, "1_carbon", "minushalf_potfiles")
+
+    ensure_dir(carbon_input_dir)
+    ensure_dir(carbon_mh_dir)
+
+    potcar_c = os.path.join(potfiles_dir, "POTCAR.c")
+
+    copy_file(potcar_c, os.path.join(carbon_mh_dir, "POTCAR.c"))
+    copy_file(potcar_c, os.path.join(carbon_input_dir, "POTCAR"))
```
Comment (ronaldorpela) [inline tests/run_tests.py]: This seems to be hard-coded for the three current tests. It would be preferable to generalize this so that adding new tests does not require modifying this logic.

File: tests/test_farm/VASP/3_nnd/minushalf.yaml [hunk 21]
```diff
@@ -0,0 +1,5 @@
+software: VASP
+vasp:
+    command: ['python', '/home/augusto/minushalf_env38/minushalf/tests/test_farm/VASP/3_nnd/nndcheck.py']
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/minushalf.yaml]: This seems to hard-code a script path specific to your local environment. It might be better to use a relative path or a configurable variable to ensure portability across different systems.

File: tests/run_tests.py (hunk 9 above)
Comment (ronaldorpela) [reply tests/run_tests.py]: I believe that this comment also applies elsewhere in this PR

File: tests/test_farm/VASP/3_nnd/nndcheck.py [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+import sys
+from pathlib import Path
+
+# ─────────────────────────────────────────────────────────────────────────────
+# CONFIGURATION
+# ─────────────────────────────────────────────────────────────────────────────
+
+# Folder containing the pre-computed VASP output folders (cut_2.59, cut_2.72, etc.)
+# Lives next to this script.
+INPUTS_DIR = Path(__file__).parent / "INPUTS"
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/nndcheck.py]: If there is no reason against it, I would recommend changing this to a local variable

File: tests/test_farm/VASP/3_nnd/nndcheck.py [hunk 23]
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env python3
+"""
+mock_vasp.py
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/nndcheck.py]: Was this perhaps an older name?

File: tests/test_farm/VASP/3_nnd/nndcheck.py [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+        for item in source_dir.iterdir():
+            dest = dest_dir / item.name
+            if item.is_file():
+                shutil.copy2(item, dest)
+            elif item.is_dir():
+                if dest.exists():
+                    shutil.rmtree(dest)
+                shutil.copytree(item, dest)
+    else:
+        print(f"WARNING: INPUTS directory not found at {source_dir.resolve()}")
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/nndcheck.py]: Would it make more sense to raise an exception here if the ref. file is not found, instead of issuing a warning?

File: tests/test_farm/VASP/2_metal/minushalf.yaml [hunk 25]
```diff
@@ -0,0 +1,5 @@
+software: VASP
+vasp:
+    command: ['python', '/home/augusto/minushalf_env38/minushalf/tests/test_farm/VASP/2_metal/metalcheck.py']
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/2_metal/minushalf.yaml]: hard-coded script path

File: tests/test_farm/VASP/2_metal/metalcheck.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+"""
+mock_vasp.py
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/2_metal/metalcheck.py]: Is there any differences with respect to nndcheck.py and sanitycheck.py?

File: tests/test_farm/VASP/1_carbon/sanitycheck.py [hunk 27]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+"""
+mock_vasp.py
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/1_carbon/sanitycheck.py]: name

File: tests/test_farm/VASP/1_carbon/minushalf.yaml [hunk 28]
```diff
@@ -0,0 +1,5 @@
+software: VASP
+vasp:
+    command: ['python', '/home/augusto/minushalf_env38/minushalf/tests/test_farm/VASP/1_carbon/sanitycheck.py']
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/1_carbon/minushalf.yaml]: hard-coded script path

**Comment by ronaldorpela** [comment]
@BACStrange : I have completed the first review round and left a few comments for improving the initial version of the test suite. I still have two additional points that I would like to discuss with you:

1.  When running pytest locally, I observed that a few existing unit tests failed. Did you also encounter that this?As far as I can tell, these failures are unrelated to the new tests introduced in this PR.
2. The test suite appears to be relatively large (72 MB vs. 116 MB for the full package). It might be worth considering reducing the k-point grid, which would also decrease the size of files such as EIGENVAL, PROCAR, and vasprun.xml. Another possibility could be to verify whether adjusting the plane-wave cutoff helps reduce these file sizes.

File: tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR (hunk 2 above)
Comment (BACStrange) [reply tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/INCAR]: As VASP would not be run directly for those tests, this tag would make no difference. Removing it anyway is case of local testing with these files.

File: tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/KPOINTS (hunk 3 above)
Comment (BACStrange) [reply tests/test_farm/VASP/1_carbon/INPUTS/cut_2.34/KPOINTS]: Updating test cases with less dense grids and smaller cutoff energies for smaller test files.

File: tests/testconfig.yml (hunk 13 above)
Comment (BACStrange) [reply tests/testconfig.yml]: I was planning to add a "setup" keyword also for the POTCAR setup and other future setups that might be needed. But it can indeed be removed in this version.

File: tests/test_farm/VASP/3_nnd/minushalf.yaml (hunk 21 above)
Comment (BACStrange) [reply tests/test_farm/VASP/3_nnd/minushalf.yaml]: The problem is, minushalf changes locations while it is running, so I couldn't figure a solution using relative paths. What I did was to make the run_tests.py script create the minushalf.yaml file with the absolute path to the test script. So, actually, for this commit the minushalf.yaml file wasn't even needed.

File: tests/run_tests.py (hunk 10 above)
Comment (BACStrange) [reply tests/run_tests.py]: I renamed the file and forgot to change its name in the comments.

File: tests/run_tests.py [hunk 29]
```diff
@@ -0,0 +1,199 @@
+"""
+run_tests.py
+------------
+Read the test registry from mhtests.yml and runs each test sequentially.
```
Comment (ronaldorpela) [inline tests/run_tests.py]: Isn't `mhtests.yml` the old name?

File: tests/run_tests.py [hunk 30]
```diff
@@ -0,0 +1,199 @@
...
+    # Report result
+    if result.returncode == 0:
+        print(f"\n[PASS] ✅ {name} passed.")
+        return True
+    else:
+        print(f"\n[FAIL] ❌ {name} failed with return code {result.returncode}.")
+        return False
+
+
+def separator():
```
Comment (ronaldorpela) [inline tests/run_tests.py]:
```suggestion
def print_separator():
```

Comment (ronaldorpela) [reply tests/run_tests.py]: In my opinion, `print_separator` seems more descriptive

File: tests/run_tests.py [hunk 31]
```diff
@@ -0,0 +1,199 @@
...
+    Args:
+        test (dict): A test entry from mhtests.yml with keys:
+                     name, description, folder, script
+    """
+    name        = test["name"]
+    description = test["description"]
+    folder      = tests_root / test["folder"]
+    script      = folder / test["script"]
+
+    print(f"\n{'='*60}")
```
Comment (ronaldorpela) [inline tests/run_tests.py]: Here, it would be better to call the corresponding function

File: tests/run_tests.py [hunk 32]
```diff
@@ -0,0 +1,199 @@
...
+    print(f"  TEST : {name}")
+    print(f"  DESC : {description}")
+    print(f"  DIR  : {folder}")
+    print(f"  SCRIPT: {script}")
+    print(f"{'='*60}\n")
```
Comment (ronaldorpela) [inline tests/run_tests.py]: Here, it would be better to call the corresponding function

File: tests/test_farm/VASP/1_carbon/sanitycheck.py (hunk 27 above)
Comment (ronaldorpela) [reply tests/test_farm/VASP/1_carbon/sanitycheck.py]: I am sorry for my previous comment: it may have not clearly described the issue.

Here, there seems to be a slight inconsistency between the documentation and the expected action suggested by the script name. The documentation describes a script that mocks a VASP calculation, while the script name suggests a sanity check. Perhaps it would be worth revisiting the intended purpose and aligning the naming and documentation accordingly.

A similar comment might also apply to the scripts used in the other test cases.

File: tests/test_farm/VASP/2_metal/INPUTS/KPOINTS [hunk 33]
```diff
@@ -0,0 +1,5 @@
+Automatic mesh
+0
+Gamma
+5 5 5
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/2_metal/INPUTS/KPOINTS]: I was wondering if a grid 5x5x5 is indeed necessary here

File: tests/test_farm/VASP/3_nnd/INPUTS/INCAR [hunk 34]
```diff
@@ -0,0 +1,15 @@
+ISTART = 0
+ICHARG = 2
+PREC   = Accurate
+EDIFF  = 1E-4
+ENCUT = 500.0
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/INPUTS/INCAR]: It might be worth revisiting the ENCUT used in this test. The 3_nnd test already produces files of around 61 MB, so reducing ENCUT could help keep the test size more manageable.

File: tests/test_farm/VASP/3_nnd/INPUTS/KPOINTS [hunk 35]
```diff
@@ -0,0 +1,5 @@
+Automatic mesh
+0
+Gamma
+15 15 15
```
Comment (ronaldorpela) [inline tests/test_farm/VASP/3_nnd/INPUTS/KPOINTS]: Do you think the 15x15x15 grid in this test might be higher than necessary? Decreasing it, might help to reduce the disk space used by the 3_nnd test

File: tests/test_farm/VASP/3_nnd/nndcheck.py (hunk 24 above)
Comment (ronaldorpela) [reply tests/test_farm/VASP/3_nnd/nndcheck.py]: It seems this point might not have been addressed in the previous review.

File: tests/run_tests.py [hunk 36]
```diff
@@ -0,0 +1,199 @@
...
+# ─────────────────────────────────────────────────────────────────────────────
+# HELPER FUNCTIONS
+# ─────────────────────────────────────────────────────────────────────────────
+
+import os
+import shutil
+
+def potcar_setup(registry, base_dir="test_farm/VASP"):
+    """
+    Copies and organizes POTCAR files for VASP test cases, based on the
```
Comment (ronaldorpela) [inline tests/run_tests.py]: According to the conventions used so far in the project, it would be more consistent to use the imperative tense in the documentation (instead of the present tense).

**Review by ronaldorpela — COMMENTED** [review]
@BACStrange,

I have completed a second review round and left a few additional comments. Apart from minor points, I am still a bit concerned about the disk usage of the third test (~61 MB), which seems relatively large. It might be worth revisiting this before moving forward, so I would suggest one more iteration before giving my approval.

Regarding your comment:

> New commit added. Answering the question about some tests failing, it was indeed the intention. The metal check and the nnd check will fail until those bugs are corrected, and just then those tests will pass. It serves to track some of the project development, as well as testing those edge cases when the project is complete.

My earlier remark was referring to the already existing unit tests that are executed when running pytest. It does not apply to the new integration tests introduced in this MR.

Would you mind taking another look at the failing unit tests? From my side, I observed some failures, which might be related to a missing installation of the atomic program, but I wanted to confirm this with you.

File: tests/run_tests.py (hunk 20 above)
Comment (BACStrange) [reply tests/run_tests.py]: Solved. Used atoms for each test now must be added in the "atoms" section on testconfig.yaml

File: tests/test_farm/VASP/2_metal/metalcheck.py (hunk 26 above)
Comment (BACStrange) [reply tests/test_farm/VASP/2_metal/metalcheck.py]: Not right now. But the test files are separated so any implementation on the tests can be done individually.

File: tests/test_farm/VASP/3_nnd/INPUTS/INCAR (hunk 34 above)
Comment (BACStrange) [reply tests/test_farm/VASP/3_nnd/INPUTS/INCAR]: Reduce the input parameters and the test folder size is now about 3MB.

File: minushalf/.libs/libatm_cGui.CVX3M6VBP5TNQYF6SPW636KVXIEUXVHJ.gfortran-win_amd64.dll
Comment (ronaldorpela) [inline minushalf/.libs/libatm_cGui.CVX3M6VBP5TNQYF6SPW636KVXIEUXVHJ.gfortran-win_amd64.dll:1]: I assume that this DLL file is not strictly required for this project.

Could you clarify the reason for including it in this PR?

**Review by ronaldorpela — APPROVED** [review]
@BACStrange 

I have reviewed this PR and consider it almost ready for merging. There is one DLL file that I assume may have been added unintentionally. 

Apart from that, everything looks good from my perspective. I am approving the PR under the assumption that this minor issue can be addressed without requiring another review round.
